superbench/benchmarks/micro_benchmarks/directx_mem_bw_performance/GPUMemRwBw.cpp

```diff
@@ -0,0 +1,314 @@
+ *		  and synchronization objects.
+ */
+void GPUMemRwBw::CreatePipeline() {
+    UINT dxgiFactoryFlags = 0;
+#if _DEBUG
+    // Enable the debug layer (requires the Graphics Tools "optional feature").
+    // NOTE: Enabling the debug layer after device creation will invalidate the active device.
+    {
+        ComPtr<ID3D12Debug> debugController;
+        if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(&debugController)))) {
+            debugController->EnableDebugLayer();
+            // Enable additional debug layers.
+            dxgiFactoryFlags |= DXGI_CREATE_FACTORY_DEBUG;
+        }
+    }
+#endif
+    ComPtr<IDXGIFactory4> factory;
+    ThrowIfFailed(CreateDXGIFactory2(dxgiFactoryFlags, IID_PPV_ARGS(&factory)));
+    ComPtr<IDXGIAdapter1> hardwareAdapter;
+    GetHardwareAdapter(factory.Get(), &hardwareAdapter);
+    ThrowIfFailed(D3D12CreateDevice(hardwareAdapter.Get(), D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&m_device)));
+    D3D12_COMMAND_QUEUE_DESC cqd3 = {};
+    cqd3.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+    ThrowIfFailed(m_device->CreateCommandQueue(&cqd3, IID_PPV_ARGS(&m_commandQueue)));
+    ThrowIfFailed(m_device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&m_commandAllocator)));
+    // Create the command list.
+    ThrowIfFailed(m_device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, m_commandAllocator.Get(), nullptr,
+                                              IID_PPV_ARGS(&m_commandList)));
+    // Create synchronization objects.
+    ThrowIfFailed(m_device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&m_fence)));
+    m_fenceValue = 1;
+    // Create an event handle to use for GPU synchronization.
+    m_eventHandle = CreateEvent(0, false, false, 0);
+}
+
+/**
+ * @brief Setup GPU pipeline resource including creating root signature, pipeline state and compile shader.
+ */
+void GPUMemRwBw::LoadAssets() {
+    // Prepare root signature, root parameter can be a table, root descriptor or root constants.
+    const int nParamter = 3;
+    CD3DX12_ROOT_PARAMETER slotRootParameter[nParamter];
+    // Bind the SRV, CBV and UAV descriptor tables to the root parameters.
+    slotRootParameter[0].InitAsShaderResourceView(0);
+    slotRootParameter[1].InitAsConstantBufferView(0);
+    slotRootParameter[2].InitAsUnorderedAccessView(0);
+    // Create the root signature.
+    // A root signature is an array of root parameters.
+    CD3DX12_ROOT_SIGNATURE_DESC rootSigDesc(nParamter, slotRootParameter, 0, nullptr, D3D12_ROOT_SIGNATURE_FLAG_NONE);
+    ComPtr<ID3DBlob> serializedRootSig = nullptr;
+    ComPtr<ID3DBlob> errorBlob = nullptr;
+    HRESULT hr = D3D12SerializeRootSignature(&rootSigDesc, D3D_ROOT_SIGNATURE_VERSION_1,
+                                             serializedRootSig.GetAddressOf(), errorBlob.GetAddressOf());
+    if (errorBlob != nullptr) {
+        ::OutputDebugStringA((char *)errorBlob->GetBufferPointer());
+        return;
+    }
+    ThrowIfFailed(hr);
```

Comment:
throwiffailed(Function) and then check if errorblob is nullptr or not?

Reply:
a little confused, if throwed, then it will not go to check errorblob?

Reply:
updated to 
```
HRESULT hr = D3D12SerializeRootSignature(&rootSigDesc, D3D_ROOT_SIGNATURE_VERSION_1,
                                             serializedRootSig.GetAddressOf(), errorBlob.GetAddressOf());
    if (hr != S_OK || errorBlob != nullptr) {
        std::cout << "Error: " << (char *)errorBlob->GetBufferPointer() << std::endl;
        throw runtime_error("Error: D3D12SerializeRootSignature failed.");
    }
```